### DISCUSSION TITLE
Callback on_train_batch_end batch on CPU

### DISCUSSION THREAD

**Discussion by import-antigravity** [Lightning-AI/pytorch-lightning#6945] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi, I have a callback with methods for on_train_batch_end and on_validation_batch_end, and I want to take the same batch and run it through another model. However, the data is on the CPU so I have to move it back to the GPU. Is there any way to avoid this redundant moving of data? It's really slowing things down.

**Comment by awaelchli** [comment]
Can you give a runnable example that demonstrates this? I can't see where Lightning would move these outputs back to CPU. 

```python
def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
    # pl_module, outputs, batch should all be on the same device
```

**Reply by import-antigravity** [reply]
Could you give me a link to the BoringModel? I can make a notebook showing what i'm talking about

**Reply by awaelchli** [reply]
https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing

**Reply by import-antigravity** [reply]
https://colab.research.google.com/drive/1axSx8lmpAhllnlcga6iE2eNgC1fjGK1b?usp=sharing

**Reply by import-antigravity** [reply]
Hopefully this works

**Reply by awaelchli** [reply]
Okay, I understand now. 
This happens because we move the batch to the device in the accelerator right before it calls training_step(). The old batch reference is still on the CPU and this will be passed to the on_train_batch_end. 

I will create an issue for this. Thanks for reporting.

**Reply by import-antigravity** [reply]
Happy to help!

**Reply by import-antigravity** [reply]
> Okay, I understand now.
> This happens because we move the batch to the device in the accelerator right before it calls training_step(). The old batch reference is still on the CPU and this will be passed to the on_train_batch_end.
> 
> I will create an issue for this. Thanks for reporting.

I don't know if you've created an issue yet but I made https://github.com/PyTorchLightning/pytorch-lightning/issues/6990 around the same time I made the OP

**Reply by awaelchli** [reply]
#6990 was probably a misunderstanding. 

I created the issue here #7377 and linked a PR. Hopefully finishing it soon fixing the benchmark test.
